pkg/tfbridge/check_failures.go

```diff
@@ -0,0 +1,350 @@
+		valuePath:   valuePath,
+		schemaMap:   p.schemaMap,
+		schemaInfos: p.schemaInfos,
+	}
+}
+
+func (p CheckFailurePath) SetElement() CheckFailurePath {
+	// Sets will be represented as lists in Pulumi; more could be done here to find the right index.
+	valuePath := p.valuePath
+	if p.isMaxItemsOne() {
+		valuePath = fmt.Sprintf("%s[?]", p.valuePath)
+	}
+	return CheckFailurePath{
+		schemaPath:  p.schemaPath.Element(),
+		valuePath:   valuePath,
+		schemaMap:   p.schemaMap,
+		schemaInfos: p.schemaInfos,
+	}
+}
+
+func (p CheckFailurePath) len() int {
+	return len(p.schemaPath)
+}
+
+func (p CheckFailurePath) topLevelPropertyKey(
+	schemaMap shim.SchemaMap,
+	schemaInfos map[string]*SchemaInfo,
+) (resource.PropertyKey, bool) {
+	if len(p.schemaPath) == 0 {
+		return "", false
+	}
+	s, ok := p.schemaPath[0].(walk.GetAttrStep)
+	if !ok {
+		return "", false
+	}
+	n := TerraformToPulumiNameV2(s.Name, schemaMap, schemaInfos)
+	return resource.PropertyKey(n), true
+}
+
+// Formats a CheckFailure to report an issue with a resource or provider configuration. This method is made public to
+// facilitate cross-package use by the bridge framework and is not intended for use by provider authors.
+func NewCheckFailure(
+	reasonType CheckFailureReason,
+	reason string,
+	pp *CheckFailurePath,
+	urn resource.URN,
+	isProvider bool,
+	configPrefix string,
+	schemaMap shim.SchemaMap,
+	schemaInfos map[string]*SchemaInfo,
+) plugin.CheckFailure {
+	if reasonType == MissingKey {
+		if isProvider {
+			return missingProviderKey(urn, configPrefix, pp, schemaMap, schemaInfos)
+		}
+		return missingRequiredKey(pp, configPrefix, schemaMap, schemaInfos)
+	}
+	if isProvider {
+		return formatProviderCheckFailure(reasonType, reason, pp, urn, configPrefix, schemaMap, schemaInfos)
+	}
+	if pp != nil && pp.valuePath != "" {
+		reason = fmt.Sprintf("%s. Examine values at '%s.%s'.", reason, urn.Name().String(), pp.valuePath)
+	}
+	return plugin.CheckFailure{
+		Reason: reason,
+		// Do not populate Property here as that changes the UX of how it displays in CLI. We may want to
+		// consider this change at some point.
+	}
+}
+
+func formatProviderCheckFailure(
+	reasonType CheckFailureReason,
+	reason string,
+	pp *CheckFailurePath,
+	urn resource.URN,
+	configPrefix string,
+	schemaMap shim.SchemaMap,
+	schemaInfos map[string]*SchemaInfo,
+) plugin.CheckFailure {
+	reason = "could not validate provider configuration: " + reason
+	if isExplicitProvider(urn) {
+		return formatExplicitProviderCheckFailure(reason, pp, urn)
+	}
+	return formatDefaultProviderCheckFailure(reasonType, reason, pp, configPrefix, schemaMap, schemaInfos)
+}
+
+func formatExplicitProviderCheckFailure(reason string, pp *CheckFailurePath, urn resource.URN) plugin.CheckFailure {
+	if pp != nil && pp.valuePath != "" && isExplicitProvider(urn) {
+		reason = fmt.Sprintf("%s. Examine values at '%s.%s'.", reason, urn.Name().String(), pp.valuePath)
+	}
+	return plugin.CheckFailure{
+		Reason: reason,
```

Comment:
Why don't we populate `Property` here? I see the comment for `formatDefaultProviderCheckFailure`, but it shouldn't apply here.

Reply:
Added comments. We can do a follow up PR, it's a change of behavior but may be a good one.